core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java

```diff
@@ -0,0 +1,543 @@
+
+        AggregationDefinition(int pageSize, ChannelDefinition... channels)
+        {
+            this.pageSize = pageSize;
+            this.channels = Arrays.stream(requireNonNull(channels, "channels is null")).collect(toImmutableList());
+        }
+    }
+
+    public static class ChannelDefinition
+    {
+        private final ColumnType columnType;
+        private final int distinctValuesCount;
+        private final int dictionaryPositions;
+        private final int distinctDictionaries;
+
+        public ChannelDefinition(ColumnType columnType, int distinctValuesCount)
+        {
+            this(columnType, distinctValuesCount, -1, -1);
+        }
+
+        public ChannelDefinition(ColumnType columnType, int distinctValuesCount, int distinctDictionaries)
+        {
+            this(columnType, distinctValuesCount, distinctValuesCount, distinctDictionaries);
+        }
+
+        public ChannelDefinition(ColumnType columnType, int distinctValuesCount, int dictionaryPositions, int distinctDictionaries)
+        {
+            this.columnType = requireNonNull(columnType, "columnType is null");
+            this.distinctValuesCount = distinctValuesCount;
+            this.dictionaryPositions = dictionaryPositions;
+            this.distinctDictionaries = distinctDictionaries;
+        }
+
+        public Block[] createBlocks(int blockCount, int positionsPerBlock, int channel, double nullChance)
+        {
+            Block[] blocks = new Block[blockCount];
+            if (dictionaryPositions == -1) { // No dictionaries
+                BlockBuilder allValues = columnType.type.createBlockBuilder(null, distinctValuesCount);
+                columnType.blockWriter.write(allValues, distinctValuesCount, channel);
+                Random r = new Random(channel);
+                for (int i = 0; i < blockCount; i++) {
+                    BlockBuilder block = columnType.type.createBlockBuilder(null, positionsPerBlock);
+                    for (int j = 0; j < positionsPerBlock; j++) {
+                        if (isNull(r, nullChance)) {
+                            block.appendNull();
+                        }
+                        else {
+                            int position = r.nextInt(distinctValuesCount);
+                            columnType.type.appendTo(allValues, position, block);
+                        }
+                    }
+                    blocks[i] = block.build();
+                }
+            }
+            else {
+                Random r = new Random(1);
+
+                int totalValueCount = Math.max(distinctValuesCount, dictionaryPositions);
+                BlockBuilder allValues = columnType.type.createBlockBuilder(null, totalValueCount);
+                columnType.blockWriter.write(allValues, totalValueCount, channel);
+                Set<Integer> usedValues = nOutOfM(r, distinctValuesCount, totalValueCount);
+
+                Block[] dictionaries = new Block[distinctDictionaries];
```

Comment:
could you add comments what is happening in each step? It's really hard to figure what is happening below with 2 level structures